src/expr/impl/src/aggregate/approx_percentile.rs

```diff
@@ -12,56 +12,220 @@
+    async fn get_result(&self, state: &AggregateState) -> Result<Datum> {
+        let state = state.downcast_ref::<State>();
+        let quantile_count = (state.count as f64 * self.quantile) as u64;
+        let mut acc_count = 0;
+        for (bucket_id, count) in state.neg_buckets.iter().rev() {
+            acc_count += count;
+            if acc_count > quantile_count {
+                // approx value = -2 * y^i / (y + 1)
+                let approx_percentile = -2.0 * self.base.powi(*bucket_id) / (self.base + 1.0);
+                let approx_percentile = ScalarImpl::Float64(approx_percentile.into());
+                return Ok(Datum::from(approx_percentile));
+            }
+        }
+        acc_count += state.zeros;
+        if acc_count > quantile_count {
+            return Ok(Datum::from(ScalarImpl::Float64(0.0.into())));
+        }
+        for (bucket_id, count) in &state.pos_buckets {
+            acc_count += count;
+            if acc_count > quantile_count {
+                // approx value = 2 * y^i / (y + 1)
+                let approx_percentile = 2.0 * self.base.powi(*bucket_id) / (self.base + 1.0);
+                let approx_percentile = ScalarImpl::Float64(approx_percentile.into());
+                return Ok(Datum::from(approx_percentile));
+            }
+        }
+        return Ok(None);
+    }
+
+    fn encode_state(&self, state: &AggregateState) -> Result<Datum> {
+        let state = state.downcast_ref::<State>();
+        let mut encoded_state = Vec::with_capacity(state.estimated_heap_size());
+        encoded_state.extend_from_slice(&state.count.to_be_bytes());
+        encoded_state.extend_from_slice(&state.zeros.to_be_bytes());
+        let neg_buckets_size =
+            state.neg_buckets.len() * (size_of::<BucketId>() + size_of::<Count>());
+        encoded_state.extend_from_slice(&neg_buckets_size.to_be_bytes());
+        for (bucket_id, count) in &state.neg_buckets {
+            encoded_state.extend_from_slice(&bucket_id.to_be_bytes());
+            encoded_state.extend_from_slice(&count.to_be_bytes());
+        }
+        for (bucket_id, count) in &state.pos_buckets {
+            encoded_state.extend_from_slice(&bucket_id.to_be_bytes());
+            encoded_state.extend_from_slice(&count.to_be_bytes());
+        }
+        let encoded_scalar = ScalarImpl::Bytea(encoded_state.into());
+        Ok(Datum::from(encoded_scalar))
     }
 
-    async fn get_result(&self, _state: &AggregateState) -> Result<Datum> {
-        todo!()
+    fn decode_state(&self, datum: Datum) -> Result<AggregateState> {
+        let mut state = State::default();
+        let Some(scalar_state) = datum else {
+            return Ok(AggregateState::Any(Box::new(state)));
+        };
+        let encoded_state: Box<[u8]> = scalar_state.into_bytea();
+        let mut cursor = 0;
+        state.count = u64::from_be_bytes(encoded_state[cursor..cursor + 8].try_into().unwrap());
+        cursor += 8;
```

Comment:
Can be simplified by using [bytes::Buf::get_u64](https://docs.rs/bytes/latest/bytes/buf/trait.Buf.html#method.get_u64)

Reply:
Fixed